plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+package avro
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+	"github.com/jeremywohl/flatten"
+	"github.com/linkedin/goavro/v2"
+	"time"
+)
+
+type Parser struct {
+	MetricName      string   `toml:"metric_name"`
+	SchemaRegistry  string   `toml:"avro_schema_registry"`
+	Schema          string   `toml:"avro_schema"`
+	Measurement     string   `toml:"avro_measurement"`
+	Tags            []string `toml:"avro_tags"`
+	Fields          []string `toml:"avro_fields"`
+	Timestamp       string   `toml:"avro_timestamp"`
+	TimestampFormat string   `toml:"avro_timestamp_format"`
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+
+	switch {
+	case p.SchemaRegistry != "":
+		schemaRegistry := NewSchemaRegistry(p.SchemaRegistry)
+		retrSchema, err := schemaRegistry.getSchema(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = retrSchema
+	default:
+		schema = p.Schema
+	}
```

Comment:
Why not use a simple `if` statement? Furthermore, do we really need to construct the registry for each parsing run or can this be done in `Init()`?

Reply:
It can certainly be an if statement.  We do, however, need to look up each schema, and each record may have a different schema--my use case consumes data from many, often dynamically-updated (so therefore the schema ID changes), schemas.  However, getSchema memoizes its results, so we only look up a given schema ID from the registry the first time we see that ID.